libc/stdlib/stdlib_strto.c

```diff
@@ -0,0 +1,1121 @@
+	TEST_ASSERT_EQUAL_PTR(str_int + 5, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+}
+
+
+TEST(stdlib_strto, empty)
+{
+	const char str[] = "";
+	char *end = NULL;
+
+	/* According to POSIX: When "No conversion could be performed", these functions may set errno to EINVAL. */
+	/* So errno in these situations can be either 0 or EINVAL nothing else */
+	errno = 0;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_DOUBLE(0, strtod(str, &end));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_EQUAL_FLOAT(0, strtof(str, &end));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_EQUAL_DOUBLE(0, strtold(str, &end));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_EQUAL_INT(0, strtol(str, &end, 10));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_EQUAL_INT(0, strtoll(str, &end, 10));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_EQUAL_INT(0, strtoul(str, &end, 10));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_EQUAL_INT(0, strtoull(str, &end, 10));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
+}
+
+
+TEST(stdlib_strto, truncate)
+{
+	const char *str_int[] = { "  123   ", "  123", "	123", "123Alma mam lkorta", "123\n\t ", "123!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~", "123\xe3\x83\x9e\xe3\x83\xaa\xe3\x82\xa2\xe3\x83\xbb" };
+	const char *str_float[] = { "  1.23  ", "  1.23", "	1.23", "1.23Alma mam lkorta", "1.23\n\t", "1.23!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~", "1.23\xe3\x83\x9e\xe3\x83\xaa\xe3\x82\xa2\xe3\x83\xbb" };
+
+	for (int i = 0; i < (sizeof(str_int) / sizeof(str_int[0])); i++) {
+		errno = 0;
+		TEST_ASSERT_EQUAL_DOUBLE(123, strtod(str_int[i], NULL));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_FLOAT(123, strtof(str_int[i], NULL));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_DOUBLE(123, strtold(str_int[i], NULL));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(1.23, strtol(str_float[i], NULL, 10));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(1.23, strtoll(str_float[i], NULL, 10));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(1.23, strtoul(str_float[i], NULL, 10));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT(1.23, strtoull(str_float[i], NULL, 10));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+	}
+}
+
+
+TEST(stdlib_strto, truncate_whitespaces)
+{
+	const char *str_int[] = { "\t123", "\v123", "\f123", "\n123", "\r123", "\r\n123", " 123" };
```

Comment:
Thanks for adding the rest of white spaces, I meant adding also a one check with all of them in one string.